### PR TITLE
Site prism

### PR DESCRIPTION
**Pull request by safeforge**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -7,4 +7,6 @@ gemspec
 gem 'rubillow', github: 'jeffdeville/rubillow'
 
 # scraper
-gem 'yasf', github: 'algonauti/yasf'
+gem 'site_prism', github: 'natritmeyer/site_prism'
```
Comment (jeffdeville) [inline Gemfile]: Why are we using unreleased versions of site_prism and poltergeist?

File: Gemfile [hunk 2]
```diff
@@ -7,4 +7,6 @@ gemspec
...
+gem 'poltergeist', github: 'teampoltergeist/poltergeist'
+gem 'capybara-screenshot'
```
Comment (jeffdeville) [inline Gemfile]: This was my fault, could we move this to a development requirement in the gemspec, or just remove the gem entirely?

File: lib/sherlock_homes.rb [hunk 3]
```diff
@@ -1,22 +1,28 @@
...
 require 'rubillow'
-require 'yasf'
+require 'site_prism'
+
+require 'sherlock_homes/version'
 
 module SherlockHomes
 
-  def self.configuration
+  def self.config
```
Comment (jeffdeville) [inline lib/sherlock_homes.rb]: +1 for brevity!  :-)

File: lib/sherlock_homes/driver.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          js_errors: false,
+          debug: false,
+          inspector: false,
+          timeout: SherlockHomes.config.timeout,
+          phantomjs_options: phantomjs_options
+        }
+
+        options.merge!(
+          logger: nil,
+          phantomjs_logger: StringIO.new,
```
Comment (jeffdeville) [inline lib/sherlock_homes/driver.rb]: Could we:
1. make the logger's configurable in a separate pull
2. Make the default logger a Ruby logger that goes to stdout

File: lib/sherlock_homes/driver.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+module SherlockHomes
+  class Driver
+
```
Comment (jeffdeville) [inline lib/sherlock_homes/driver.rb]: Another future PR:  Can we make the driver not be global?  When we get to workflows, we might want to use different drivers for different sites.  (I saw there was a capybara driver that just used RestClient or something like that behind the scenes). Might be nice to use that for non-js sites. (Created a pivotal story in the icebox for this)

File: lib/sherlock_homes/scraper.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+module SherlockHomes
+  class Scraper
+
+    def self.restart_phantomjs
```
Comment (jeffdeville) [inline lib/sherlock_homes/scraper.rb]: Under what circumstances is this used?  Does it just clear cookies?

File: sherlock_homes.gemspec [hunk 7]
```diff
@@ -21,6 +21,7 @@ Gem::Specification.new do |spec|
 
   # Runtime
   spec.add_dependency 'geocoder', '~> 1.2'
+  spec.add_dependency 'activesupport', '~> 4.2'
```
Comment (jeffdeville) [inline sherlock_homes.gemspec]: Is this high a level of activesupport required?  Is it just being used for autoload?  Let's try to lower this requirement as far as possible for older Rails users.

File: spec/lib/trulia_spec.rb [hunk 8]
```diff
@@ -1,60 +1,56 @@
...
-RSpec.describe SherlockHomes::Trulia, sauce: true do
+RSpec.describe SherlockHomes::Trulia do
 
-  Given(:scraper) do
-    subject.base_url 'http://www.trulia.com/homes/Pennsylvania/Allentown/sold/1263178-2490-Riverbend-Rd-Allentown-PA-18103'
-    subject
+  When do
+    subject.class.set_url 'http://www.trulia.com/homes/Pennsylvania/Allentown/sold/1263178-2490-Riverbend-Rd-Allentown-PA-18103'
+    subject.load
+    subject.wait_for_estimates
```
Comment (jeffdeville) [inline spec/lib/trulia_spec.rb]: Should wait_for_estimates be done automatically in load, or perhaps better yet, added in dynamically when accessing estimate data?  (Trying to simplify the api for the user)

File: spec/lib/redfin_spec.rb [hunk 9]
```diff
@@ -1,15 +1,72 @@
...
-RSpec.describe SherlockHomes::Redfin, sauce: true do
+RSpec.describe SherlockHomes::Redfin, skip_ci: true do
 
-  Given(:scraper) do
-    subject.base_url 'https://www.redfin.com/PA/Allentown/2354-S-Cedar-Crest-Blvd-18103/home/59034427'
-    subject
+  When do
+    subject.class.set_url 'https://www.redfin.com/PA/Allentown/2354-S-Cedar-Crest-Blvd-18103/home/59034427'
+    subject.load
+    subject.wait_for_contact_box
```
Comment (jeffdeville) [inline spec/lib/redfin_spec.rb]: What is: wait_for_contact_box?  I'm assuming it's an ajax thing that we need to wait for before finishing pulling all of the data. I'd prefer to bake this into the scraper though.

**Comment by jeffdeville** [comment]
I had a few requests, but this is a massive amount of code.  Great work @safeforge!

File: lib/sherlock_homes/scraper.rb (hunk 6 above)
Comment (safeforge) [inline lib/sherlock_homes/scraper.rb]: when a multiple sites are scraped phantom need to be restarted. e.g.

```ruby
google_link = Google.search("#{address} site:redfin.com")
Scraper.restart_phantomjs
redfin_result = Redfin.find(google_link)
Scraper.restart_phantomjs
google_link = Google.search("#{address} site:trulia.com")
Scraper.restart_phantomjs
trulia_result = Trulia.find(address)
```

File: sherlock_homes.gemspec (hunk 7 above)
Comment (safeforge) [inline sherlock_homes.gemspec]: at this moment we using only the `core_ext` I mean `"string".present?` `"mulit string".underscore`

File: spec/lib/trulia_spec.rb (hunk 8 above)
Comment (safeforge) [inline spec/lib/trulia_spec.rb]: do you mean a static method called for example `find(url)` ?

File: lib/sherlock_homes/driver.rb (hunk 4 above)
Comment (safeforge) [inline lib/sherlock_homes/driver.rb]: `phantomjs_logger` should be a class with public `puts` method, default `Logger` has this method private and probably not is compatible. I'll try and send you a feedback

File: lib/sherlock_homes/scraper.rb (hunk 6 above)
Comment (jeffdeville) [inline lib/sherlock_homes/scraper.rb]: Why is this the case?

**Pull request merged**